panic: Out Of Memory Quota!

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)
panic: Out Of Memory Quota![conn_id=3971954969015222677]

goroutine 2161 [running]:
[github.com/pingcap/tidb/util/memory.(*PanicOnExceed).Action](http://github.com/pingcap/tidb/util/memory.(*PanicOnExceed).Action)(0xc78c02ba00, 0xc001246c80)
        /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/util/memory/action.go:148](http://github.com/pingcap/tidb/util/memory/action.go:148) +0x11f
[github.com/pingcap/tidb/util/memory.(*Tracker).Consume.func2](http://github.com/pingcap/tidb/util/memory.(*Tracker).Consume.func2)(0x45440c0?, 0xc0002b5980?)
        /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/util/memory/tracker.go:449](http://github.com/pingcap/tidb/util/memory/tracker.go:449) +0xff
[github.com/pingcap/tidb/util/memory.(*Tracker).Consume](http://github.com/pingcap/tidb/util/memory.(*Tracker).Consume)(0xc2087b3620?, 0x1000)
        /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/util/memory/tracker.go:461](http://github.com/pingcap/tidb/util/memory/tracker.go:461) +0x2c2
[github.com/pingcap/tidb/util/memory.(*Tracker).ReplaceBytesUsed](http://github.com/pingcap/tidb/util/memory.(*Tracker).ReplaceBytesUsed)(0xc778af1540, 0xc0b4be5ef0?)
        /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/util/memory/tracker.go:743](http://github.com/pingcap/tidb/util/memory/tracker.go:743) +0x3c
[github.com/pingcap/tidb/session.(*session).SetMemoryFootprintChangeHook.func1(0xc42b905388?)](http://github.com/pingcap/tidb/session.(*session).SetMemoryFootprintChangeHook.func1(0xc42b905388?))
        /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/session/session.go:4096](http://github.com/pingcap/tidb/session/session.go:4096) +0xc5
[github.com/tikv/client-go/v2/internal/unionstore.(*MemDB).SetMemoryFootprintChangeHook.func1()](http://github.com/tikv/client-go/v2/internal/unionstore.(*MemDB).SetMemoryFootprintChangeHook.func1())
        /go/pkg/mod/[github.com/tikv/client-go/v2@v2.0.6-0.20230209044022-95ebf6a86301/internal/unionstore/memdb.go:875](http://github.com/tikv/client-go/v2@v2.0.6-0.20230209044022-95ebf6a86301/internal/unionstore/memdb.go:875) +0x2c
[github.com/tikv/client-go/v2/internal/unionstore.(*memdbArena).onMemChange(0xc42b9053c0?)](http://github.com/tikv/client-go/v2/internal/unionstore.(*memdbArena).onMemChange(0xc42b9053c0?))
        /go/pkg/mod/[github.com/tikv/client-go/v2@v2.0.6-0.20230209044022-95ebf6a86301/internal/unionstore/memdb_arena.go:141](http://github.com/tikv/client-go/v2@v2.0.6-0.20230209044022-95ebf6a86301/internal/unionstore/memdb_arena.go:141) +0x33
[github.com/tikv/client-go/v2/internal/unionstore.(*MemDB).Cleanup](http://github.com/tikv/client-go/v2/internal/unionstore.(*MemDB).Cleanup)(0xc68ccfcd20, 0x1)
        /go/pkg/mod/[github.com/tikv/client-go/v2@v2.0.6-0.20230209044022-95ebf6a86301/internal/unionstore/memdb.go:158](http://github.com/tikv/client-go/v2@v2.0.6-0.20230209044022-95ebf6a86301/internal/unionstore/memdb.go:158) +0x185
[github.com/pingcap/tidb/store/driver/txn.(*memBuffer).Cleanup](http://github.com/pingcap/tidb/store/driver/txn.(*memBuffer).Cleanup)(0x38c057c?, 0xc68ccfcd20?)
        /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/store/driver/txn/unionstore_driver.go:74](http://github.com/pingcap/tidb/store/driver/txn/unionstore_driver.go:74) +0x1c
[github.com/pingcap/tidb/session.(*LazyTxn).changeToInvalid(0xc008fd4c90)](http://github.com/pingcap/tidb/session.(*LazyTxn).changeToInvalid(0xc008fd4c90))
        /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/session/txn.go:315](http://github.com/pingcap/tidb/session/txn.go:315) +0x4f
[github.com/pingcap/tidb/session.(*LazyTxn).reset(0x0?)](http://github.com/pingcap/tidb/session.(*LazyTxn).reset(0x0?))
        /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/session/txn.go:555](http://github.com/pingcap/tidb/session/txn.go:555) +0x28
[github.com/pingcap/tidb/session.(*LazyTxn).Rollback(0xc008fd4c90)](http://github.com/pingcap/tidb/session.(*LazyTxn).Rollback(0xc008fd4c90))
        /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/session/txn.go:436](http://github.com/pingcap/tidb/session/txn.go:436) +0xb4
[github.com/pingcap/tidb/session.(*session).RollbackTxn](http://github.com/pingcap/tidb/session.(*session).RollbackTxn)(0xc008fd4c80, {0x5297910?, 0xc10d6d9ad0?})
        /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/session/session.go:1112](http://github.com/pingcap/tidb/session/session.go:1112) +0xda
[github.com/pingcap/tidb/session.(*session).Close(0xc008fd4c80)](http://github.com/pingcap/tidb/session.(*session).Close(0xc008fd4c80))
        /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/session/session.go:2564](http://github.com/pingcap/tidb/session/session.go:2564) +0x265
[github.com/pingcap/tidb/server.(*TiDBContext).Close(0xc007ee7fc8)](http://github.com/pingcap/tidb/server.(*TiDBContext).Close(0xc007ee7fc8))
        /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/server/driver_tidb.go:268](http://github.com/pingcap/tidb/server/driver_tidb.go:268) +0xb7
[github.com/pingcap/tidb/server.closeConn](http://github.com/pingcap/tidb/server.closeConn)(0xc0077e5080, 0xc001968e40?)
        /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/server/conn.go:356](http://github.com/pingcap/tidb/server/conn.go:356) +0x1dc
[github.com/pingcap/tidb/server.(*clientConn).Close(0xc0077e5080)](http://github.com/pingcap/tidb/server.(*clientConn).Close(0xc0077e5080))
        /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/server/conn.go:340](http://github.com/pingcap/tidb/server/conn.go:340) +0x89
[github.com/pingcap/tidb/server.(*Server).onConn.func3()](http://github.com/pingcap/tidb/server.(*Server).onConn.func3())
        /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/server/server.go:601](http://github.com/pingcap/tidb/server/server.go:601) +0x2f
[github.com/pingcap/tidb/server.(*Server).onConn](http://github.com/pingcap/tidb/server.(*Server).onConn)(0xc0046bc800, 0xc0077e5080)
        /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/server/server.go:641](http://github.com/pingcap/tidb/server/server.go:641) +0xbb4
created by [github.com/pingcap/tidb/server.(*Server).startNetworkListener](http://github.com/pingcap/tidb/server.(*Server).startNetworkListener)
        /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/server/server.go:482](http://github.com/pingcap/tidb/server/server.go:482) +0x5ca
### 4. What is your TiDB version? (Required)
6.6.0
<!-- Paste the output of SELECT tidb_version() -->

global memory controller killed a wrong sql.
``` sh
2022232:[2023/02/16 11:25:58.424 +08:00] [WARN] [servermemorylimit.go:126] ["global memory controller tries to kill the top1 memory consumer"] [connID=3971954969015222677] ["sql digest"=6d0590fcb088a7be00d3c625cec85e43f71d7f46e581b844b83be2dbf5b54da0] ["sql text"=] [tidb_server_memory_limit=27487790640] ["heap inuse"=32395190272] ["sql memory usage"=37856]
2022233:[2023/02/16 11:25:58.427 +08:00] [WARN] [tracker.go:458] ["global memory controller, NeedKill signal is received successfully"] [connID=3971954969015222677]
2022234:[2023/02/16 11:25:58.427 +08:00] [WARN] [expensivequery.go:118] ["memory exceeds quota"] [cost_time=0.006575794s] [conn_id=3971954969015222677] [user=root] [database=longrun_tpcc] [index_names="[new_order:PRIMARY]"] [txn_start_ts=439489123622846523] [mem_max="38177 Bytes (37.3 KB)"] [sql=]
2022235:[2023/02/16 11:25:58.427 +08:00] [ERROR] [coprocessor.go:1073] ["copIteratorWork meet panic"] [r="\"Out Of Memory Quota![conn_id=3971954969015222677]\""] ["stack trace"="github.com/pingcap/tidb/store/copr.(*copIteratorWorker).handleTask.func1\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/store/copr/coprocessor.go:1075\nruntime.gopanic\n\t/usr/local/go/src/runtime/panic.go:884\ngithub.com/pingcap/tidb/util/memory.(*PanicOnExceed).Action\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/util/memory/action.go:148\ngithub.com/pingcap/tidb/util/memory.(*Tracker).Consume.func2\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/util/memory/tracker.go:449\ngithub.com/pingcap/tidb/util/memory.(*Tracker).Consume\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/util/memory/tracker.go:461\ngithub.com/pingcap/tidb/store/copr.(*copIteratorWorker).sendToRespCh\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/store/copr/coprocessor.go:963\ngithub.com/pingcap/tidb/store/copr.(*copIteratorWorker).handleCopResponse\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/store/copr/coprocessor.go:1342\ngithub.com/pingcap/tidb/store/copr.(*copIteratorWorker).handleCopPagingResult\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/store/copr/coprocessor.go:1244\ngithub.com/pingcap/tidb/store/copr.(*copIteratorWorker).handleTaskOnce\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/store/copr/coprocessor.go:1183\ngithub.com/pingcap/tidb/store/copr.(*copIteratorWorker).handleTask\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/store/copr/coprocessor.go:1086\ngithub.com/pingcap/tidb/store/copr.(*copIteratorWorker).run\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/store/copr/coprocessor.go:782"]
2022236:[2023/02/16 11:25:58.428 +08:00] [ERROR] [conn.go:1064] ["connection running loop panic"] [conn=3971954969015222677] [lastSQL="SELECT no_o_id FROM new_order WHERE no_w_id = ? AND no_d_id = ? ORDER BY no_o_id ASC LIMIT 1 FOR UPDATE [arguments: (253, 5)]"] [err="Out Of Memory Quota![conn_id=3971954969015222677]"] [stack="github.com/pingcap/tidb/server.(*clientConn).Run.func1\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/server/conn.go:1067\nruntime.gopanic\n\t/usr/local/go/src/runtime/panic.go:884\ngithub.com/pingcap/tidb/util/memory.(*PanicOnExceed).Action\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/util/memory/action.go:148\ngithub.com/pingcap/tidb/util/memory.(*Tracker).Consume.func2\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/util/memory/tracker.go:449\ngithub.com/pingcap/tidb/util/memory.(*Tracker).Consume\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/util/memory/tracker.go:461\ngithub.com/pingcap/tidb/util/memory.(*Tracker).ReplaceBytesUsed\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/util/memory/tracker.go:743\ngithub.com/pingcap/tidb/session.(*session).SetMemoryFootprintChangeHook.func1\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/session/session.go:4096\ngithub.com/tikv/client-go/v2/internal/unionstore.(*MemDB).SetMemoryFootprintChangeHook.func1\n\t/go/pkg/mod/github.com/tikv/client-go/v2@v2.0.6-0.20230209044022-95ebf6a86301/internal/unionstore/memdb.go:875\ngithub.com/tikv/client-go/v2/internal/unionstore.(*memdbArena).onMemChange\n\t/go/pkg/mod/github.com/tikv/client-go/v2@v2.0.6-0.20230209044022-95ebf6a86301/internal/unionstore/memdb_arena.go:141\ngithub.com/tikv/client-go/v2/internal/unionstore.(*MemDB).Cleanup\n\t/go/pkg/mod/github.com/tikv/client-go/v2@v2.0.6-0.20230209044022-95ebf6a86301/internal/unionstore/memdb.go:158\ngithub.com/pingcap/tidb/store/driver/txn.(*memBuffer).Cleanup\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/store/driver/txn/unionstore_driver.go:74\ngithub.com/pingcap/tidb/session.(*LazyTxn).cleanupStmtBuf\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/session/txn.go:151\ngithub.com/pingcap/tidb/session.(*LazyTxn).cleanup\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/session/txn.go:559\ngithub.com/pingcap/tidb/session.(*session).StmtRollback\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/session/txn.go:730\ngithub.com/pingcap/tidb/session.finishStmt\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/session/tidb.go:243\ngithub.com/pingcap/tidb/session.runStmt\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/session/session.go:2361\ngithub.com/pingcap/tidb/session.(*session).ExecuteStmt\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/session/session.go:2199\ngithub.com/pingcap/tidb/server.(*TiDBContext).ExecuteStmt\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/server/driver_tidb.go:247\ngithub.com/pingcap/tidb/server.(*clientConn).executePreparedStmtAndWriteResult\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/server/conn_stmt.go:266\ngithub.com/pingcap/tidb/server.(*clientConn).executePlanCacheStmt\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/server/conn_stmt.go:224\ngithub.com/pingcap/tidb/server.(*clientConn).handleStmtExecute\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/server/conn_stmt.go:216\ngithub.com/pingcap/tidb/server.(*clientConn).dispatch\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/server/conn.go:1407\ngithub.com/pingcap/tidb/server.(*clientConn).Run\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/server/conn.go:1129\ngithub.com/pingcap/tidb/server.(*Server).onConn\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/server/server.go:625"]
```

--- a/executor/aggregate.go
+++ b/executor/aggregate.go
@@ -254,7 +254,6 @@ func (e *HashAggExec) Close() error {
 		e.partialResultMap = nil
 		if e.memTracker != nil {
 			e.memTracker.ReplaceBytesUsed(0)
-			e.memTracker.Detach()
 		}
 		if e.listInDisk != nil {
 			firstErr = e.listInDisk.Close()
@@ -290,9 +289,6 @@ func (e *HashAggExec) Close() error {
 			e.memTracker.ReplaceBytesUsed(0)
 		}
 	}
-	if e.memTracker != nil {
-		e.memTracker.Detach()
-	}
 	return e.baseExecutor.Close()
 }
 
@@ -309,7 +305,11 @@ func (e *HashAggExec) Open(ctx context.Context) error {
 	}
 	e.prepared = false
 
-	e.memTracker = memory.NewTracker(e.id, -1)
+	if e.memTracker != nil {
+		e.memTracker.Reset()
+	} else {
+		e.memTracker = memory.NewTracker(e.id, -1)
+	}
 	if e.ctx.GetSessionVars().TrackAggregateMemoryUsage {
 		e.memTracker.AttachTo(e.ctx.GetSessionVars().StmtCtx.MemTracker)
 	}
@@ -1292,8 +1292,13 @@ func (e *StreamAggExec) Open(ctx context.Context) error {
 		e.memUsageOfInitialPartialResult += memDelta
 	}
 
-	// bytesLimit <= 0 means no limit, for now we just track the memory footprint
-	e.memTracker = memory.NewTracker(e.id, -1)
+	if e.memTracker != nil {
+		e.memTracker.Reset()
+	} else {
+
+		// bytesLimit <= 0 means no limit, for now we just track the memory footprint
+		e.memTracker = memory.NewTracker(e.id, -1)
+	}
 	if e.ctx.GetSessionVars().TrackAggregateMemoryUsage {
 		e.memTracker.AttachTo(e.ctx.GetSessionVars().StmtCtx.MemTracker)
 	}
@@ -1309,9 +1314,6 @@ func (e *StreamAggExec) Close() error {
 		e.childResult = nil
 	}
 	e.groupChecker.reset()
-	if e.memTracker != nil {
-		e.memTracker.Detach()
-	}
 	return e.baseExecutor.Close()
 }
 

--- a/executor/aggregate.go
+++ b/executor/aggregate.go
@@ -254,6 +254,7 @@ func (e *HashAggExec) Close() error {
 		e.partialResultMap = nil
 		if e.memTracker != nil {
 			e.memTracker.ReplaceBytesUsed(0)
+			e.memTracker.Detach()
 		}
 		if e.listInDisk != nil {
 			firstErr = e.listInDisk.Close()
@@ -288,6 +289,9 @@ func (e *HashAggExec) Close() error {
 		if e.memTracker != nil {
 			e.memTracker.ReplaceBytesUsed(0)
 		}
+	}
+	if e.memTracker != nil {
+		e.memTracker.Detach()
 	}
 	return e.baseExecutor.Close()
 }
@@ -1305,6 +1309,9 @@ func (e *StreamAggExec) Close() error {
 		e.childResult = nil
 	}
 	e.groupChecker.reset()
+	if e.memTracker != nil {
+		e.memTracker.Detach()
+	}
 	return e.baseExecutor.Close()
 }
 

--- a/executor/aggregate.go
+++ b/executor/aggregate.go
@@ -1295,7 +1295,6 @@ func (e *StreamAggExec) Open(ctx context.Context) error {
 	if e.memTracker != nil {
 		e.memTracker.Reset()
 	} else {
-
 		// bytesLimit <= 0 means no limit, for now we just track the memory footprint
 		e.memTracker = memory.NewTracker(e.id, -1)
 	}

--- a/executor/cte.go
+++ b/executor/cte.go
@@ -112,7 +112,11 @@ func (e *CTEExec) Open(ctx context.Context) (err error) {
 		return err
 	}
 
-	e.memTracker = memory.NewTracker(e.id, -1)
+	if e.memTracker != nil {
+		e.memTracker.Reset()
+	} else {
+		e.memTracker = memory.NewTracker(e.id, -1)
+	}
 	e.diskTracker = disk.NewTracker(e.id, -1)
 	e.memTracker.AttachTo(e.ctx.GetSessionVars().StmtCtx.MemTracker)
 	e.diskTracker.AttachTo(e.ctx.GetSessionVars().StmtCtx.DiskTracker)
@@ -221,10 +225,6 @@ func (e *CTEExec) Close() (err error) {
 		if err = e.reopenTbls(); err != nil {
 			return err
 		}
-	}
-
-	if e.memTracker != nil {
-		e.memTracker.Detach()
 	}
 	return e.baseExecutor.Close()
 }

--- a/executor/cte.go
+++ b/executor/cte.go
@@ -223,6 +223,9 @@ func (e *CTEExec) Close() (err error) {
 		}
 	}
 
+	if e.memTracker != nil {
+		e.memTracker.Detach()
+	}
 	return e.baseExecutor.Close()
 }
 

--- a/executor/distsql.go
+++ b/executor/distsql.go
@@ -229,9 +229,6 @@ func (e *IndexReaderExecutor) Close() (err error) {
 		return nil
 	}
 	e.ctx.StoreQueryFeedback(e.feedback)
-	if e.memTracker != nil {
-		e.memTracker.Detach()
-	}
 	return err
 }
 
@@ -318,7 +315,12 @@ func (e *IndexReaderExecutor) open(ctx context.Context, kvRanges []kv.KeyRange) 
 		return nil
 	}
 
-	e.memTracker = memory.NewTracker(e.id, -1)
+	if e.memTracker != nil {
+		e.memTracker.Reset()
+	} else {
+
+		e.memTracker = memory.NewTracker(e.id, -1)
+	}
 	e.memTracker.AttachTo(e.ctx.GetSessionVars().StmtCtx.MemTracker)
 	slices.SortFunc(kvRanges, func(i, j kv.KeyRange) bool {
 		return bytes.Compare(i.StartKey, j.StartKey) < 0

--- a/executor/distsql.go
+++ b/executor/distsql.go
@@ -229,6 +229,9 @@ func (e *IndexReaderExecutor) Close() (err error) {
 		return nil
 	}
 	e.ctx.StoreQueryFeedback(e.feedback)
+	if e.memTracker != nil {
+		e.memTracker.Detach()
+	}
 	return err
 }
 

--- a/executor/distsql.go
+++ b/executor/distsql.go
@@ -318,7 +318,6 @@ func (e *IndexReaderExecutor) open(ctx context.Context, kvRanges []kv.KeyRange) 
 	if e.memTracker != nil {
 		e.memTracker.Reset()
 	} else {
-
 		e.memTracker = memory.NewTracker(e.id, -1)
 	}
 	e.memTracker.AttachTo(e.ctx.GetSessionVars().StmtCtx.MemTracker)

--- a/executor/executor.go
+++ b/executor/executor.go
@@ -1629,6 +1629,9 @@ func (e *SelectionExec) Close() error {
 		e.childResult = nil
 	}
 	e.selected = nil
+	if e.memTracker != nil {
+		e.memTracker.Detach()
+	}
 	return e.baseExecutor.Close()
 }
 

--- a/executor/executor.go
+++ b/executor/executor.go
@@ -1609,7 +1609,11 @@ func (e *SelectionExec) Open(ctx context.Context) error {
 }
 
 func (e *SelectionExec) open(ctx context.Context) error {
-	e.memTracker = memory.NewTracker(e.id, -1)
+	if e.memTracker != nil {
+		e.memTracker.Reset()
+	} else {
+		e.memTracker = memory.NewTracker(e.id, -1)
+	}
 	e.memTracker.AttachTo(e.ctx.GetSessionVars().StmtCtx.MemTracker)
 	e.childResult = tryNewCacheChunk(e.children[0])
 	e.memTracker.Consume(e.childResult.MemoryUsage())
@@ -1629,9 +1633,6 @@ func (e *SelectionExec) Close() error {
 		e.childResult = nil
 	}
 	e.selected = nil
-	if e.memTracker != nil {
-		e.memTracker.Detach()
-	}
 	return e.baseExecutor.Close()
 }
 

--- a/executor/index_lookup_hash_join.go
+++ b/executor/index_lookup_hash_join.go
@@ -307,6 +307,9 @@ func (e *IndexNestedLoopHashJoin) Close() error {
 	e.joinChkResourceCh = nil
 	e.finished.Store(false)
 	e.prepared = false
+	if e.memTracker != nil {
+		e.memTracker.Detach()
+	}
 	return e.baseExecutor.Close()
 }
 

--- a/executor/index_lookup_hash_join.go
+++ b/executor/index_lookup_hash_join.go
@@ -128,7 +128,11 @@ func (e *IndexNestedLoopHashJoin) Open(ctx context.Context) error {
 	if err != nil {
 		return err
 	}
-	e.memTracker = memory.NewTracker(e.id, -1)
+	if e.memTracker != nil {
+		e.memTracker.Reset()
+	} else {
+		e.memTracker = memory.NewTracker(e.id, -1)
+	}
 	e.memTracker.AttachTo(e.ctx.GetSessionVars().StmtCtx.MemTracker)
 	e.cancelFunc = nil
 	e.innerPtrBytes = make([][]byte, 0, 8)
@@ -307,9 +311,6 @@ func (e *IndexNestedLoopHashJoin) Close() error {
 	e.joinChkResourceCh = nil
 	e.finished.Store(false)
 	e.prepared = false
-	if e.memTracker != nil {
-		e.memTracker.Detach()
-	}
 	return e.baseExecutor.Close()
 }
 

--- a/executor/index_merge_reader.go
+++ b/executor/index_merge_reader.go
@@ -174,7 +174,11 @@ func (e *IndexMergeReaderExecutor) Open(ctx context.Context) (err error) {
 	}
 	e.finished = make(chan struct{})
 	e.resultCh = make(chan *indexMergeTableTask, atomic.LoadInt32(&LookupTableTaskChannelSize))
-	e.memTracker = memory.NewTracker(e.id, -1)
+	if e.memTracker != nil {
+		e.memTracker.Reset()
+	} else {
+		e.memTracker = memory.NewTracker(e.id, -1)
+	}
 	e.memTracker.AttachTo(e.ctx.GetSessionVars().StmtCtx.MemTracker)
 	return nil
 }
@@ -795,9 +799,6 @@ func (e *IndexMergeReaderExecutor) Close() error {
 	e.processWorkerWg.Wait()
 	e.finished = nil
 	e.workerStarted = false
-	if e.memTracker != nil {
-		e.memTracker = nil
-	}
 	// TODO: how to store e.feedbacks
 	return nil
 }

--- a/executor/index_merge_reader.go
+++ b/executor/index_merge_reader.go
@@ -795,6 +795,9 @@ func (e *IndexMergeReaderExecutor) Close() error {
 	e.processWorkerWg.Wait()
 	e.finished = nil
 	e.workerStarted = false
+	if e.memTracker != nil {
+		e.memTracker = nil
+	}
 	// TODO: how to store e.feedbacks
 	return nil
 }

--- a/executor/join.go
+++ b/executor/join.go
@@ -187,6 +187,9 @@ func (e *HashJoinExec) Close() error {
 	if e.stats != nil {
 		defer e.ctx.GetSessionVars().StmtCtx.RuntimeStatsColl.RegisterStats(e.id, e.stats)
 	}
+	if e.memTracker != nil {
+		e.memTracker.Detach()
+	}
 	err := e.baseExecutor.Close()
 	return err
 }

--- a/executor/join.go
+++ b/executor/join.go
@@ -187,9 +187,6 @@ func (e *HashJoinExec) Close() error {
 	if e.stats != nil {
 		defer e.ctx.GetSessionVars().StmtCtx.RuntimeStatsColl.RegisterStats(e.id, e.stats)
 	}
-	if e.memTracker != nil {
-		e.memTracker.Detach()
-	}
 	err := e.baseExecutor.Close()
 	return err
 }
@@ -202,7 +199,11 @@ func (e *HashJoinExec) Open(ctx context.Context) error {
 		return err
 	}
 	e.prepared = false
-	e.hashJoinCtx.memTracker = memory.NewTracker(e.id, -1)
+	if e.hashJoinCtx.memTracker != nil {
+		e.hashJoinCtx.memTracker.Reset()
+	} else {
+		e.hashJoinCtx.memTracker = memory.NewTracker(e.id, -1)
+	}
 	e.hashJoinCtx.memTracker.AttachTo(e.ctx.GetSessionVars().StmtCtx.MemTracker)
 
 	e.diskTracker = disk.NewTracker(e.id, -1)

--- a/executor/projection.go
+++ b/executor/projection.go
@@ -98,7 +98,11 @@ func (e *ProjectionExec) open(_ context.Context) error {
 	e.prepared = false
 	e.parentReqRows = int64(e.maxChunkSize)
 
-	e.memTracker = memory.NewTracker(e.id, -1)
+	if e.memTracker != nil {
+		e.memTracker.Reset()
+	} else {
+		e.memTracker = memory.NewTracker(e.id, -1)
+	}
 	e.memTracker.AttachTo(e.ctx.GetSessionVars().StmtCtx.MemTracker)
 
 	// For now a Projection can not be executed vectorially only because it
@@ -326,9 +330,6 @@ func (e *ProjectionExec) Close() error {
 			runtimeStats.SetConcurrencyInfo(execdetails.NewConcurrencyInfo("Concurrency", int(e.numWorkers)))
 		}
 		e.ctx.GetSessionVars().StmtCtx.RuntimeStatsColl.RegisterStats(e.id, runtimeStats)
-	}
-	if e.memTracker != nil {
-		e.memTracker.Detach()
 	}
 	return e.baseExecutor.Close()
 }

--- a/executor/projection.go
+++ b/executor/projection.go
@@ -327,6 +327,9 @@ func (e *ProjectionExec) Close() error {
 		}
 		e.ctx.GetSessionVars().StmtCtx.RuntimeStatsColl.RegisterStats(e.id, runtimeStats)
 	}
+	if e.memTracker != nil {
+		e.memTracker.Detach()
+	}
 	return e.baseExecutor.Close()
 }
 

--- a/executor/table_reader.go
+++ b/executor/table_reader.go
@@ -272,6 +272,9 @@ func (e *TableReaderExecutor) Close() error {
 		return nil
 	}
 	e.ctx.StoreQueryFeedback(e.feedback)
+	if e.memTracker != nil {
+		e.memTracker.Detach()
+	}
 	return err
 }
 

--- a/executor/table_reader.go
+++ b/executor/table_reader.go
@@ -142,7 +142,11 @@ func (e *TableReaderExecutor) Open(ctx context.Context) error {
 		time.Sleep(time.Millisecond * time.Duration(ms))
 	})
 
-	e.memTracker = memory.NewTracker(e.id, -1)
+	if e.memTracker != nil {
+		e.memTracker.Reset()
+	} else {
+		e.memTracker = memory.NewTracker(e.id, -1)
+	}
 	e.memTracker.AttachTo(e.ctx.GetSessionVars().StmtCtx.MemTracker)
 
 	var err error
@@ -272,9 +276,6 @@ func (e *TableReaderExecutor) Close() error {
 		return nil
 	}
 	e.ctx.StoreQueryFeedback(e.feedback)
-	if e.memTracker != nil {
-		e.memTracker.Detach()
-	}
 	return err
 }
 

--- a/util/memory/tracker.go
+++ b/util/memory/tracker.go
@@ -739,8 +739,14 @@ func (t *Tracker) DetachFromGlobalTracker() {
 
 // ReplaceBytesUsed replace bytesConsume for the tracker
 func (t *Tracker) ReplaceBytesUsed(bytes int64) {
-	t.Consume(-t.BytesConsumed())
-	t.Consume(bytes)
+	t.Consume(bytes - t.BytesConsumed())
+}
+
+// Reset detach the tracker from the old parent and clear the old children. The label and byteLimit would not be reset.
+func (t *Tracker) Reset() {
+	t.Detach()
+	t.ReplaceBytesUsed(0)
+	t.mu.children = nil
 }
 
 func (t *Tracker) getParent() *Tracker {

